sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,294 @@
+  uint32_t read_ptr_val = read_addr & VALUE_MASK;
+
+  // This represents the case where the valid data of the fifo is "inclusive",
+  // i.e., the buffer looks like (where a / represents valid data):
+  // [ /////       ]
+  //   ^    ^
+  //   read write
+  if (write_phase == read_phase) {
+    return write_ptr_val - read_ptr_val;
+  }
+
+  // This represents the case where the valid data of the fifo is "exclusive",
+  // i.e., the buffer looks like (where a / represents valid data):
+  // [/      //////]
+  //   ^     ^
+  //   write read
+  return fifo_len - (read_ptr_val - write_ptr_val);
+}
+
+/**
+ * Increments the FIFO address pointed to by |addr| by |increment|.
+ *
+ * If the pointer would wrap around the FIFO buffer, it gets wrapped and the
+ * phase bit is flipped, since this causes the FIFO to switch between
+ * "inclusive" and "exclusive" modes (see |fifo_depth()| above).
+ *
+ * A fifo pointer can be approximated by a 12-bit sign-and-magnitude integer.
+ */
+static inline void fifo_increment(uint32_t *addr, uint32_t increment,
+                                  size_t fifo_len) {
+  uint32_t phase_bit = *addr & PHASE_MASK;
+  uint32_t inc_ptr = (*addr & VALUE_MASK) + increment;
+
+  // If we would overflow, wrap and flip the sign bit.
+  if (inc_ptr >= fifo_len) {
+    inc_ptr = (inc_ptr - fifo_len) & VALUE_MASK;
+    phase_bit = ~phase_bit & PHASE_MASK;
+  }
+
+  *addr = inc_ptr | phase_bit;
+}
+
+#define IS_WORD_ALIGNED(addr) (((uintptr_t)(addr)&0b11) == 0)
+
+size_t dif_spi_device_send(const dif_spi_device_t *spi, const void *data,
+                           size_t len) {
+  uint32_t txf_ptr = reg32_read(spi->base_addr, SPI_DEVICE_TXF_PTR_REG);
+  uint32_t txf_write_addr = (txf_ptr >> SPI_DEVICE_TXF_PTR_WPTR_OFFSET) &
+                            SPI_DEVICE_TXF_PTR_WPTR_MASK;
+  uint32_t txf_read_addr = (txf_ptr >> SPI_DEVICE_TXF_PTR_RPTR_OFFSET) &
+                           SPI_DEVICE_TXF_PTR_RPTR_MASK;
+
+  size_t fifo_in_use_len =
+      fifo_depth(txf_write_addr, txf_read_addr, SPI_DEVICE_TXF_SIZE);
+  size_t space_left = SPI_DEVICE_TXF_SIZE - fifo_in_use_len - sizeof(uint32_t);
+
+  size_t message_len;
+  if (len < space_left) {
+    message_len = len;
+  } else {
+    message_len = space_left;
+  }
+
+  size_t bytes_left = message_len;
+
+  // First, bring txf_write into word-alignment, so we can do full-word writes.
+  if (!IS_WORD_ALIGNED(txf_write_addr)) {
+    // Pull off the sub-word indexing in txf_write_addr.
+    size_t misalignment = txf_write_addr & (sizeof(uint32_t) - 1);
+    size_t needed_bytes = sizeof(uint32_t) - misalignment;
+    if (needed_bytes > bytes_left) {
+      needed_bytes = bytes_left;
+    }
+
+    uint32_t value =
+        reg32_read(spi->base_addr, SPI_DEVICE_TXF_OFFSET(txf_write_addr));
+    // NOTE: we go uint32_t * -> char * -> void *, because arithmetic on void*
+    // is a GCC extension.
+    __builtin_memcpy((void *)(((char *)&value) + misalignment), data,
```

Comment:
You shouldn't need the cast to `void*` here, as all pointers are assignment-compatible with `void*`, including `char*` - was this done in response to a particular compiler warning?

I would also much prefer you used the `memcpy` that has already been implemented, which the compiler will most likely recognise and optimise anyway.

Reply:
I don't want to pull in a header with memcpy in it, since I want to keep the headers that DIFs pull in to some kind of whitelist (e.g. common.h is 100% not allowed imo). If we declare "this specific memory operations header is ok" then I'll use that.

Reply:
(Also, I got rid of the void* casts.)